<a href="https://colab.research.google.com/github/Sukrut11/Everlytics-Internship-Test/blob/main/Everlytics_Machine_Learning_Test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

####Importing the Libraries

In [2]:
import numpy as np 
import pandas as pd 
import os 
import matplotlib.pyplot as plt 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.linear_model import Lasso
from sklearn.feature_selection import SelectFromModel

In [3]:
df = pd.read_csv("/content/Phishing Data - Phishing Data.csv")
df.head()

having_IP_Address  URL_Length  Shortining_Service  having_At_Symbol  \
0                  1           1                   0                 0   
1                  0           1                   0                 0   
2                  0           0                   0                 0   
3                  0           0                   0                 0   
4                  0           0                   1                 0   

   double_slash_redirecting  Prefix_Suffix  having_Sub_Domain  SSLfinal_State  \
0                         1             -1                 -1              -1   
1                         0             -1                  0               1   
2                         0             -1                 -1              -1   
3                         0             -1                 -1              -1   
4                         0             -1                  1               1   

   Domain_registeration_length  Favicon  ...  popUpWidnow  Iframe  \
0                            0        0  ...            0       0   
1                            0        0  ...            0       0   
2                            0        0  ...            0       0   
3                            1        0  ...            0       0   
4                            0        0  ...            1       0   

   age_of_domain  DNSRecord  web_traffic  Page_Rank  Google_Index  \
0             -1          1           -1         -1             0   
1             -1          1            0         -1             0   
2              0          1            1         -1             0   
3             -1          1            1         -1             0   
4             -1          1            0         -1             0   

   Links_pointing_to_page  Statistical_report  Result  
0                       1                   1       1  
1                       1                   0       1  
2                       0                   1       1  
3                      -1                   0       1  
4                       1                   0       0  

[5 rows x 31 columns]

In [4]:
df.shape

(2456, 31)

In [5]:
df["Result"].value_counts()

0    1362
1    1094
Name: Result, dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2456 entries, 0 to 2455
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   having_IP_Address            2456 non-null   int64
 1   URL_Length                   2456 non-null   int64
 2   Shortining_Service           2456 non-null   int64
 3   having_At_Symbol             2456 non-null   int64
 4   double_slash_redirecting     2456 non-null   int64
 5   Prefix_Suffix                2456 non-null   int64
 6   having_Sub_Domain            2456 non-null   int64
 7   SSLfinal_State               2456 non-null   int64
 8   Domain_registeration_length  2456 non-null   int64
 9   Favicon                      2456 non-null   int64
 10  port                         2456 non-null   int64
 11  HTTPS_token                  2456 non-null   int64
 12  Request_URL                  2456 non-null   int64
 13  URL_of_Anchor                2456 non-null   int

In [7]:
df.describe()

having_IP_Address   URL_Length  Shortining_Service  having_At_Symbol  \
count        2456.000000  2456.000000         2456.000000       2456.000000   
mean            0.113192    -0.649837            0.122964          0.054560   
std             0.316892     0.752690            0.328463          0.227166   
min             0.000000    -1.000000            0.000000          0.000000   
25%             0.000000    -1.000000            0.000000          0.000000   
50%             0.000000    -1.000000            0.000000          0.000000   
75%             0.000000    -1.000000            0.000000          0.000000   
max             1.000000     1.000000            1.000000          1.000000   

       double_slash_redirecting  Prefix_Suffix  having_Sub_Domain  \
count               2456.000000    2456.000000        2456.000000   
mean                   0.125407      -0.254886          -0.185668   
std                    0.331247       0.676170           0.802068   
min                    0.000000      -1.000000          -1.000000   
25%                    0.000000      -1.000000          -1.000000   
50%                    0.000000       0.000000           0.000000   
75%                    0.000000       0.000000           0.000000   
max                    1.000000       1.000000           1.000000   

       SSLfinal_State  Domain_registeration_length      Favicon  ...  \
count     2456.000000                  2456.000000  2456.000000  ...   
mean         0.255700                     0.018730     0.189739  ...   
std          0.912332                     0.798455     0.392175  ...   
min         -1.000000                    -1.000000     0.000000  ...   
25%         -1.000000                    -1.000000     0.000000  ...   
50%          1.000000                     0.000000     0.000000  ...   
75%          1.000000                     1.000000     0.000000  ...   
max          1.000000                     1.000000     1.000000  ...   

       popUpWidnow       Iframe  age_of_domain    DNSRecord  web_traffic  \
count  2456.000000  2456.000000    2456.000000  2456.000000  2456.000000   
mean      0.196254     0.092020      -0.003257     0.536645     0.304560   
std       0.397244     0.289112       0.939726     0.498757     0.834146   
min       0.000000     0.000000      -1.000000     0.000000    -1.000000   
25%       0.000000     0.000000      -1.000000     0.000000     0.000000   
50%       0.000000     0.000000       0.000000     1.000000     1.000000   
75%       0.000000     0.000000       1.000000     1.000000     1.000000   
max       1.000000     1.000000       1.000000     1.000000     1.000000   

         Page_Rank  Google_Index  Links_pointing_to_page  Statistical_report  \
count  2456.000000   2456.000000             2456.000000         2456.000000   
mean     -0.540717      0.139658                0.344463            0.179153   
std       0.757832      0.346702                0.568911            0.383559   
min      -1.000000      0.000000               -1.000000            0.000000   
25%      -1.000000      0.000000                0.000000            0.000000   
50%      -1.000000      0.000000                0.000000            0.000000   
75%       0.000000      0.000000                1.000000            0.000000   
max       1.000000      1.000000                1.000000            1.000000   

            Result  
count  2456.000000  
mean      0.445440  
std       0.497115  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000  

[8 rows x 31 columns]

#### Data Splitting and Feature Selection

In [8]:
x = df.drop(columns = "Result", axis = 1)
y = df["Result"]

In [9]:
feature_sel_model = SelectFromModel(Lasso(alpha=0.001, random_state=0)) # remember to set the seed, the random state in this function
feature_sel_model.fit(x, y)

SelectFromModel(estimator=Lasso(alpha=0.001, random_state=0))

In [10]:
feature_sel_model.get_support() # True- Important Feature and False- Not Important Feature

array([False,  True, False,  True,  True,  True,  True,  True,  True,
       False,  True,  True,  True,  True,  True, False,  True, False,
       False, False, False,  True, False,  True,  True,  True,  True,
        True,  True, False])

In [11]:
selected_feat = x.columns[(feature_sel_model.get_support())]

# let's print some stats
print('total features: {}'.format((x.shape[1])))
print('selected features: {}'.format(len(selected_feat)))
print('features with coefficients shrank to zero: {}'.format(
np.sum(feature_sel_model.estimator_.coef_ == 0)))

total features: 30
selected features: 20
features with coefficients shrank to zero: 10


In [12]:
selected_feat

Index(['URL_Length', 'having_At_Symbol', 'double_slash_redirecting',
       'Prefix_Suffix', 'having_Sub_Domain', 'SSLfinal_State',
       'Domain_registeration_length', 'port', 'HTTPS_token', 'Request_URL',
       'URL_of_Anchor', 'Links_in_tags', 'Submitting_to_email', 'popUpWidnow',
       'age_of_domain', 'DNSRecord', 'web_traffic', 'Page_Rank',
       'Google_Index', 'Links_pointing_to_page'],
      dtype='object')

In [13]:
X= df[selected_feat]
X.head()

URL_Length  having_At_Symbol  double_slash_redirecting  Prefix_Suffix  \
0           1                 0                         1             -1   
1           1                 0                         0             -1   
2           0                 0                         0             -1   
3           0                 0                         0             -1   
4           0                 0                         0             -1   

   having_Sub_Domain  SSLfinal_State  Domain_registeration_length  port  \
0                 -1              -1                            0     0   
1                  0               1                            0     0   
2                 -1              -1                            0     0   
3                 -1              -1                            1     0   
4                  1               1                            0     0   

   HTTPS_token  Request_URL  URL_of_Anchor  Links_in_tags  \
0            1            1             -1              1   
1            1            1              0             -1   
2            1            1              0             -1   
3            1           -1              0              0   
4            0            1              0              0   

   Submitting_to_email  popUpWidnow  age_of_domain  DNSRecord  web_traffic  \
0                    1            0             -1          1           -1   
1                    0            0             -1          1            0   
2                    1            0              0          1            1   
3                    0            0             -1          1            1   
4                    0            1             -1          1            0   

   Page_Rank  Google_Index  Links_pointing_to_page  
0         -1             0                       1  
1         -1             0                       1  
2         -1             0                       0  
3         -1             0                      -1  
4         -1             0                       1

In [14]:
Y = df["Result"]
Y.head()

0    1
1    1
2    1
3    1
4    0
Name: Result, dtype: int64

In [15]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.1, stratify=y, random_state = 40)

In [16]:
print(X_train.shape, Y_train.shape, X_test.shape, Y_test.shape)

(2210, 20) (2210,) (246, 20) (246,)


#### Applying Logistic Regression

In [17]:
model = LogisticRegression()
model.fit(X_train, Y_train)

LogisticRegression()

In [18]:
Y_pred_train = model.predict(X_train)
training_data_accuracy = accuracy_score(Y_pred_train, Y_train)

In [19]:
print("The accuracy of the training data is ", training_data_accuracy)

The accuracy of the training data is  0.9447963800904977


In [20]:
Y_pred_test = model.predict(X_test)
test_data_accuracy = accuracy_score(Y_pred_test, Y_test)

In [21]:
print("The accuracy of testing data is ", test_data_accuracy*100)

The accuracy of testing data is  95.9349593495935


In [47]:
# Testing the accuracy of our Logistic Regression model manually on dataset input
input_data = X[756 : 757]
input_data_to_array = np.asarray(input_data)
print(input_data_to_array)

[[-1  0  0 -1  0 -1  1  1  0 -1 -1  0  1  1  1  1  0 -1  1  1]]


In [48]:
input_data_to_array_reshape = input_data_to_array.reshape(1, -1) # 1 means one row and -1 means multiple columns

In [49]:
input_data_to_array_reshape

array([[-1,  0,  0, -1,  0, -1,  1,  1,  0, -1, -1,  0,  1,  1,  1,  1,
         0, -1,  1,  1]])

In [50]:
predictions = model.predict(input_data_to_array_reshape)
print(predictions)
if (predictions[0] == 0):
  print("The website is not Phishy")
else:
  print("The website is Phishy!!")

[1]
The website is Phishy!!


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  "X does not have valid feature names, but"


In [51]:
print(Y[756])

1


#### Applying Decision Tree Classifier

In [22]:
# Using Decision Trees Classifier
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier(criterion='gini', max_depth=20, random_state=0)
dtc.fit(X_train, Y_train)

DecisionTreeClassifier(max_depth=20, random_state=0)

In [23]:
y_predict = dtc.predict(X_test)
print(y_predict)

[0 1 0 0 1 0 0 0 0 0 0 1 1 0 1 0 1 1 0 1 0 0 0 1 1 0 1 1 0 0 0 0 0 0 0 0 0
 1 0 1 1 1 1 1 0 0 1 1 1 1 0 0 0 0 1 1 1 0 0 1 0 1 0 1 1 0 0 0 1 0 1 1 0 1
 1 0 1 0 1 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 1 0 0 0 0 0 1 1 1 0 1 1 0 1 0 1
 1 0 1 0 0 1 1 1 1 0 1 1 1 0 1 1 1 1 0 1 0 0 0 0 0 0 1 1 1 1 0 0 0 0 0 1 1
 1 1 1 1 0 1 0 0 1 1 1 0 0 0 0 1 0 1 0 1 0 0 0 1 0 1 1 1 0 0 1 1 1 1 0 1 1
 0 1 1 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 1 1 0 0 0 1 1 1
 0 1 0 0 0 1 0 1 0 1 0 0 0 0 0 0 1 1 0 1 1 0 1 0]


In [24]:
print("Model Accuracy is",accuracy_score(Y_test, y_predict) )

Model Accuracy is 0.983739837398374


In [25]:
print('Decision Tree Classifier', np.abs(accuracy_score(Y_test, y_predict))*100)

Decision Tree Classifier 98.3739837398374


In [35]:
# Testing the accuracy of our Decision Tree model manually on dataset input
input_data = X[45 : 46]
input_data_to_array = np.asarray(input_data)
print(input_data_to_array)

[[-1  0  0 -1  1  1  0  0  0  1  1  0  1  0  1  0  1  1  0  0]]


In [36]:
input_data_to_array_reshape = input_data_to_array.reshape(1, -1) # 1 means one row and -1 means multiple columns

In [37]:
input_data_to_array_reshape

array([[-1,  0,  0, -1,  1,  1,  0,  0,  0,  1,  1,  0,  1,  0,  1,  0,
         1,  1,  0,  0]])

In [41]:
predictions = model.predict(input_data_to_array_reshape)
print(predictions)
if (predictions[0] == 0):
  print("The website is not Phishy")
else:
  print("The website is Phishy!!")

[0]
The website is not Phishy


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  "X does not have valid feature names, but"


In [42]:
print(Y[45])

0


#### Applying Random Forest Classifier

In [26]:
from sklearn.ensemble import RandomForestClassifier
rf= RandomForestClassifier(n_estimators = 20,criterion='entropy',random_state = 0)

In [27]:
rf.fit(X_train, Y_train)

RandomForestClassifier(criterion='entropy', n_estimators=20, random_state=0)

In [28]:
pred = rf.predict(X_test)

In [29]:
score = rf.score(X_test, Y_test)
print('Randomforest Classifier', np.abs(score)*100)

Randomforest Classifier 97.5609756097561


In [30]:
# Testing the accuracy of our Random Forest model manually on dataset input
input_data = X[2 : 3]
input_data_to_array = np.asarray(input_data)
print(input_data_to_array)

[[ 0  0  0 -1 -1 -1  0  0  1  1  0 -1  1  0  0  1  1 -1  0  0]]


In [31]:
input_data_to_array_reshape = input_data_to_array.reshape(1, -1)

In [32]:
input_data_to_array_reshape

array([[ 0,  0,  0, -1, -1, -1,  0,  0,  1,  1,  0, -1,  1,  0,  0,  1,
         1, -1,  0,  0]])

In [33]:
predictions = model.predict(input_data_to_array_reshape)
print(predictions)
if (predictions[0] == 0):
  print("The website is not Phishy")
else:
  print("The website is Phishy!!")

[1]
The website is Phishy!!


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  "X does not have valid feature names, but"


In [34]:
print(Y[2])

1


Conclusion -
1. I got 95 % accuracy from Logistic Regression model on test data.
2. I got 97 % accuracy from RandomForest Classifier model on test data.
3. I got 98 % accuracy from Decision Tree Classifier model on test data.